File: solution/295/solution.cpp
class MedianFinder {
public:
    /** initialize your data structure here. */
    MedianFinder() {}

    void addNum(int num) {
        if(big_queue.empty()){
            big_queue.push(num);
            return;
        }
        if(big_queue.size()==small_queue.size()){
            if(num<big_queue.top()){big_queue.push(num);}
            else{small_queue.push(num);}
        }else if(big_queue.size()>small_queue.size()){
            if(num<big_queue.top()){
                small_queue.push(big_queue.top());
                big_queue.pop();
                big_queue.push(num);
            }
            else{small_queue.push(num);}
        }else{
            if(num>small_queue.top()){
                big_queue.push(small_queue.top());
                small_queue.pop();
                small_queue.push(num);
            }
            else{big_queue.push(num);}
        }

    }

    double findMedian() {
        if(big_queue.size()==small_queue.size()){return (big_queue.top() + small_queue.top())/2;}
        else if(big_queue.size()>small_queue.size()){return big_queue.top();}
        else{return small_queue.top();}
    }
private:
    std::priority_queue<double> big_queue;
    std::priority_queue<double, std::vector<double>, std::greater<double> > small_queue;
};

File: solution/295/solution.py
import heapq

class MedianFinder(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.big_queue = []
        self.small_queue = []
        heapq.heapify(self.big_queue)
        heapq.heapify(self.small_queue)


    def addNum(self, num):
        """
        :type num: int
        :rtype: void
        """
        if not self.big_queue:
            heapq.heappush(self.big_queue, -num)
            return
        if len(self.big_queue) == len(self.small_queue):
            if num < -self.big_queue[0]:
                heapq.heappush(self.big_queue, -num)
            else:
                heapq.heappush(self.small_queue, num)
        elif len(self.big_queue) > len(self.small_queue):
            if num < -self.big_queue[0]:
                heapq.heappush(self.small_queue, -self.big_queue[0])
                heapq.heapreplace(self.big_queue, -num)
            else:
                heapq.heappush(self.small_queue, num)
        else:
            if num > self.small_queue[0]:
                heapq.heappush(self.big_queue, -self.small_queue[0])
                heapq.heapreplace(self.small_queue, num)
            else:
                heapq.heappush(self.big_queue, -num)


    def findMedian(self):
        """
        :rtype: float
        """
        if len(self.big_queue) == len(self.small_queue):
            return (self.small_queue[0]-self.big_queue[0])/2.0
        elif len(self.big_queue) > len(self.small_queue):
            return -self.big_queue[0]
        else:
            return self.small_queue[0]